README
This is an extension for typo3's new form-wizard (introduced in v4.6).
A vcard containing details sent by form is attached to the mail.

Use as form_postprocessor:

 * Log into backend.
 * Select site to create form.
 * Insert new form-plugin.
 * Create form
 	+ either by using the form wizard,
 	+ or by using typoscript (see example below).
 * Set postProcessor.X to 'tx_vcardmail'

postProcessor {
	1 = tx_vcardmail
	1 {
		recipientEmail = foo@bar.com
		senderEmail = bar@foo.com
		# version can either be 2.1 or 3.0. Defaults to 2.1.
		version = 2.1
		properties {

	# A vCard must only have one of the following properties each.
	# Tha declaration numbered highest overrides earlier declarations.
	# Set .field to the name of the field.
			1 = FULLNAME
			1.field = name
			2 = TITLE
			2.field = title-field
			3 = LASTNAME
			3.field = lastname-field
			4 = FIRSTNAME
			4.field = firstname-field
			5 = ADDITIONALNAMES
			5.field = additionalnames-field
			6 = NICKNAME
			6.field = nickname-field
			7 = NAMEPREFIX
			7.field = nameprefix-field
			8 = NAMESUFFIX
			8.field = namesuffix-field
			9 = BIRTHDAY
			9.field = birthday-field
			# format	date-conf	The format of the submitted data.\nSee the PHP-manual (date) for the codes, or datatype "date-conf" in the TYPO3 document TSref.	d-m-Y
			9.format = d-m-Y
			10 = ORGANIZATION
			10.field = organization-field
			11 = DEPARTMENT
			11.field = department-field
			12 = SUBDEPARTMENT
			12.field = subdepartment-field
			13 = ROLE
			13.field = role-field
			# if REVISION is not set current date and time will be used.
			14 = REVISION
			14.field = revision-field
			# format	strftime-conf	The format of the submitted data.\nSee the PHP-manual (date) for the codes, or datatype "date-conf" in the TYPO3 document TSref.	d-m-Y  H:i:s
			14.format = d-m-Y H:i:s
			16 = GENDER
			16.field = gender-field
			17 = GEOLOCATION
			17.field = location-field
			18 = MAILER
			18.field = mailer-field
			# This property is specified in a manner consistent with ISO 8601. The property value is a signed numeric indicating the number of hours and possibly minutes from UTC. Time zones east of UTC are positive numbers. Time zones west of UTC are negative numbers.
			19 = TIMEZONE
			19.field = timezone-field
			20 = UID
			20.field = uid-field
			
	# A vCard can have multiple entries of the following properties.
	# Set .field to the name of the field.
	# .type can have none to many entries.
			21 = ADDRESS
			21 {
				city = city-field
				state = state-field
				zip = zipcode-field
				street = street-field
				country = country-field
				extendedaddress = extendedaddress-field
				postofficeaddress = postofficeaddress-field
				type {
					# allowed types for an address are:
					1 = DOM
					2 = INTL
					3 = POSTAL
					4 = PARCEL
					5 = HOME
					6 = WORK
				}
			}
			22 = EMAIL
			22 {
				field = email
				type {
					# allowed types for an e-mail-address are:
					1 = INTERNET
					2 = WORK
					3 = HOME
					4 = OTHER
					5 = PREF
				}
			}
			23 = PHONENUMBER
			23 {
				field = telephone-field
				type {
					# allowed types for a telephone number are:
					1 = WORK
					2 = HOME
					3 = OTHER
					4 = PREF
					5 = VOICE
					6 = FAX
					7 = MSG
					8 = CELL
					9 = CAR
				}
			}
			24 = URL
			24 {
				field = url-field
				type {
					# allowed types for an URL are:
					1 = WORK
					2 = HOME
					3 = OTHER
					4 = PREF
				}
			}
			25 = INSTANTMESSENGER
			25 {
				field = im-field
				# .messenger can be any out of:
				# x-AIM, X-ICQ, X-JABBER, X-MSN, X-YAHOO, X-SKYPE, X-SKYPE-USERNAME, X-GADUGADU
				messenger = X-JABBER
				type {
					# allowed types for an IM-entry are:
					1 = WORK
					2 = HOME
					3 = OTHER
					4 = PREF
				}
			}
		}
	}
}



ChangeLog
2013-06-15 Franz Geiger  <mail_at_fx-g.de>

	* Introduce namespaces. Make vcardmail ready for TYPO3 6

2012-03-05 Franz Geiger  <mail_at_fx-g.de>

	* initial code generated with kickstarter


ext_emconf.php
<?php

########################################################################
# Extension Manager/Repository config file for ext "vcardmail".
#
# Auto generated 06-03-2012 01:34
#
# Manual updates:
# Only the data in the array - everything else is removed by next
# writing. "version" and "dependencies" must not be touched!
########################################################################

$EM_CONF[$_EXTKEY] = array(
	'title' => 'VCard Mailform Postprocessor',
	'description' => 'Extends core form\'s mail postprocessor by attaching a vCard featuring the submitted contact details to mails sent by the mailform.',
	'category' => 'fe',
	'author' => 'Franz Geiger',
	'author_email' => 'mail@fx-g.de',
	'shy' => '',
	'dependencies' => 'form',
	'conflicts' => '',
	'priority' => '',
	'module' => '',
	'state' => 'beta',
	'internal' => '',
	'uploadfolder' => 0,
	'createDirs' => '',
	'modify_tables' => '',
	'clearCacheOnLoad' => 0,
	'lockType' => '',
	'author_company' => '',
	'version' => '0.1.1',
	'docPath' => 'doc',
	'constraints' => array(
		'depends' => array(
			'form' => '6.0.0-0.0.0',
			'typo3' => '6.0.0-6.0.99'
		),
		'conflicts' => array(
		),
		'suggests' => array(
		),
	),
	'_md5_values_when_last_written' => 'a:7:{s:9:"ChangeLog";s:4:"b42e";s:6:"README";s:4:"ab58";s:22:"class.tx_vcardmail.php";s:4:"47cc";s:16:"ext_autoload.php";s:4:"90fd";s:12:"ext_icon.gif";s:4:"1bdc";s:48:"libraries/zendvcard/class.tx_vcardmail_vcard.php";s:4:"750c";s:64:"libraries/zendvcard/Vcard/class.tx_vcardmail_vcard_generator.php";s:4:"e767";}',
	'suggests' => array(
	),
);
?>


class.tx_vcardmail.php
<?php

/***************************************************************
 *  Copyright notice
 *
 *  (c) 2012-2013 Franz Geiger <mail_at_fx-g.de>
 *  All rights reserved
 *
 *  This script is part of the TYPO3 project. The TYPO3 project is
 *  free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  The GNU General Public License can be found at
 *  http://www.gnu.org/copyleft/gpl.html.
 *
 *  This script is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  This copyright notice MUST APPEAR in all copies of the script!
 ***************************************************************/

/**
 * A postprocessor extending the mail postprocessor by attaching entered
 * contact details as a vCard.
 *
 * @author Franz Geiger <mail_at_fx-g.de>
 */
class tx_vcardmail extends \TYPO3\CMS\Form\PostProcess\MailPostProcessor implements \TYPO3\CMS\Form\PostProcess\PostProcessorInterface {

	/** 
	 * Version of the vCard. Must be either 2.1 or 3.0. Defaults to 2.1
	 *
	 * @var string
	 */
	protected $version = tx_vcardmail_vcard::VERSION21;
	/**
	 * Filename of the vCard file. Defaults to vcard.vcf
	 *
	 * @var string
	 */
	protected $filename = 'vcard.vcf';
	/**
	 * Mimetype of the vCard file. Defaults to text/x-vcard
	 *
	 * @var string
	 */
	protected $mimetype = 'text/x-vcard';
	/**
	 * Representation of the vCard object.
	 *
	 * @var tx_vcardmail_vcard
	 */
	protected $vcard;

	/**
	 * Output format of birthday date.
	 *
	 * @var string
	 */
	const BDAY_FORMAT = 'Y-m-d';
	/**
	 * Output format of revision date. ISO8601.
	 *
	 * @var string
	 */
	const REV_FORMAT = DateTime::ISO8601;
	/**
	 * Default input format of birthday date.
	 *
	 * @var string
	 */
	const DEFAULT_BDAY_INPUT_FORMAT = 'd-m-Y';
	/**
	 * Default input format of revision date.
	 *
	 * @var string
	 */
	const DEFAULT_REV_INPUT_FORMAT = 'd-m-Y H:i:s';


	/**
	 * Add attachments when uploaded. Overrides the inherited function in
	 * tx_form_System_Postprocessor_Mail by adding a call to
	 * $this->addVcardAttachment().
	 *
	 * @override
	 * @return void
	 */
	protected function addAttachmentsFromForm() {
		parent::addAttachmentsFromForm();

		$values = $this->requestHandler->getByMethod();
		$this->addVcardAttachment($values);
	}

	/**
	 * Attach vcard based on typoscript fields in
	 * $typoScript['postprocessor'].x.properties.
	 *
	 * @param array $submittedValues
	 * @return void
	 */
	protected function addVcardAttachment($submittedValues) {
		$this->vcard = \TYPO3\CMS\Core\Utility\GeneralUtility::makeInstance('tx_vcardmail_vcard');

		// version can be eiter 2.1 (default) or 3.0
		if ($this->typoScript['version'] == tx_vcardmail_vcard::VERSION30) {
			$this->version = tx_vcardmail_vcard::VERSION30;
		}

		if ($this->typoscript['fileName']) {
			$this->filename = $this->typoScript['fileName'];
		}

		if ($this->typoscript['mimeType']) {
			$this->mimetype = $this->typoScript['mimeType'];
		}

		$this->generateVcardFromTyposcript(
			$this->typoScript['properties.'],
			$submittedValues
		);
		$vcardString = tx_vcardmail_vcard_generator::generate($this->vcard, $this->version);

		$this->mailMessage->attach(
			Swift_Attachment::newInstance(
				$vcardString,
				$this->filename,
				$this->mimetype
			)
		);
	}

	/**
	 * Create vcard based on typoscript fields in
	 * $typoScript['postprocessors'].x.properties.
	 *
	 * @param array $typoscript	Array of properties defined in typoScript
	 * @param array $submittedValues
	 */
	protected function generateVcardFromTyposcript($typoscript, $submittedValues) {

		$keys =  \TYPO3\CMS\Core\TypoScript\TemplateService::sortedKeyList($typoscript);

		foreach ($keys as $propertyKey) {
			if (!intval($propertyKey) || strpos($propertyKey, '.') !== FALSE) {
				continue;
			}

			// Properties are defined by typoscript
			// (e.g. postProcessor.1.properties.1 = EMAIL)
			$property = ucfirst(strtolower($typoscript[$propertyKey]));

			if ($property == 'Address' ||
				$property == 'Phonenumber' ||
				$property == 'Email' ||
				$property == 'Url' ||
				$property == 'Instantmessenger'
			) {
				$this->addMultiProperty($property, $typoscript[$propertyKey . '.'], $submittedValues);

			} elseif ($property == 'Birthday' || $property == 'Revision') {
				$this->setDateProperty($property, $typoscript[$propertyKey . '.'], $submittedValues);

			} else {
				// if it's not one of the properties above: just try calling this:
				$this->setSingleProperty($property, $typoscript[$propertyKey . '.'], $submittedValues);
			}
		}

		// If REVISION is not set yet: set to current time
		if ($this->vcard->__get('revision') === NULL) {

			$revisionDate = \TYPO3\CMS\Core\Utility\GeneralUtility::makeInstance('DateTime');
			$this->vcard->setRevision($revisionDate->format(self::REV_FORMAT));
		}
	}

	/**
	 * Extracts types from typoscript.
	 *
	 * @param array $types numbered array of types
	 */
	static protected function getPropertyTypes($typeArray) {

		$types = array();
		// Types are mapped to this entry by typoscript
		// (e.g. postProcessor.1.properties.1.type.1 = WORK)
		$typeKeys = \TYPO3\CMS\Core\TypoScript\TemplateService::sortedKeyList($typeArray);

		foreach ($typeKeys as $typeKey) {
			if (!intval($typeKey) || strpos($typeKey, '.') !== FALSE) {
				continue;
			}

			$types[] = strtoupper($typeArray[$typeKey]);
		}
		return $types;
	}

	/**
	 * Adds another instance of $property to the vcard.
	 *
	 * @param string $property
	 * @param array $typoscript TS-configuration of this property
	 * @param array $submittedValues
	 */
	protected function addMultiProperty($property, $typoscript, $submittedValues) {

		$entryArray[0] = array();

		// (e.g. postProcessor.1.properties.1.zip = zip-fieldname
		//  or   postProcessor.1.properties.2.field = phone-fieldname)

		foreach ($typoscript as $key => $field) {

			if ($key == 'type.') {
				$entryArray[0]['type'] = self::getPropertyTypes($typoscript[$key]);

			} elseif ($key == 'messenger') {
				// addInstantmessenger() takes the type of messenger as an additional parameter
				// (e.g. postProcessor.1.properties.1.messenger = X-JABBER)
				$entryArray[0][$key] = $typoscript[$key];

			} elseif (isset($submittedValues[$field])) {
				if ($key == 'field') {
					$entryArray[0]['value'] = $submittedValues[$field];

				} else {
					$entryArray[0][$key] = $submittedValues[$field];
				}

			} elseif ($key == 'field' && $property != 'Address') {
				// all this properties except address require 'value'
				return;
			}
		}
		$this->vcard->{'add' . $property}($entryArray);
	}

	/**
	 * Sets $property in vcard as a date.
	 *
	 * @param string $property
	 * @param array $typoscript TS-configuration of this property
	 * @param array $submittedValues
	 */
	protected function setDateProperty($property, $typoscript, $submittedValues) {

		if (!isset($submittedValues[$typoscript['field']])) {
			return;
		}

		$value = $submittedValues[$typoscript['field']];
		$outputFormat;

		if ($property == 'Birthday') {
			$outputFormat = self::BDAY_FORMAT;
		} else {
			$outputFormat = self::REV_FORMAT;
		}

		$format = $typoscript['format'];
		if ($format !== NULL) {
			if ($property == 'Birthday') {
				$format = self::DEFAULT_BDAY_INPUT_FORMAT;
			} else {
				$format = self::DEFAULT_REV_INPUT_FORMAT;
			}
		}

		$date = DateTime::createFromFormat($format, $value);

		$this->vcard->{'set' . $property}($date->format($outputFormat));
	}

	/**
	 * Sets $property in the vcard.
	 *
	 * @param string $property
	 * @param array $typoscript TS-configuration of this property
	 * @param array $submittedValues
	 */
	protected function setSingleProperty($property, $typoscript, $submittedValues) {

		if (!isset($submittedValues[$typoscript['field']])) {
			return;
		}

		try {
			$this->vcard->{'set' . $property}($submittedValues[$typoscript['field']]);
		} catch (Exception $e) {}
	}
}
?>


libraries/zendvcard/Vcard/class.tx_vcardmail_vcard_generator.php
<?php

/***************************************************************
 * Copyright 2010 Thomas Schaaf <Thomaschaaf@gmail.com>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * Changes by: Franz Geiger <mail_at_fx-g.de>
 * Original code is available at:
 * http://code.google.com/p/zendvcard/
 ***************************************************************/
/**
 * A generator of a Vcard.
 *
 * @author Thomas Schaaf <Thomaschaaf@gmail.com>
 */
class tx_vcardmail_vcard_generator
{

	protected static function _renderTimezone($vcardObject)
	{
		if ($vcardObject->timezone)
		{
			return "TZ:" . $vcardObject->timezone . tx_vcardmail_vcard::LINEBREAK;
		}
	}

	protected static function _renderMailer($vcardObject)
	{
		if ($vcardObject->mailer)
		{
			return "MAILER:" . $vcardObject->mailer . tx_vcardmail_vcard::LINEBREAK;
		}
	}

	protected static function _renderGeolocation($vcardObject)
	{
		if ($vcardObject->geolocation)
		{
			return "GEO:" . $vcardObject->geolocation . tx_vcardmail_vcard::LINEBREAK;
		}
	}

	protected static function _renderBirthday($vcardObject)
	{
		if ($vcardObject->birthday)
		{
			return "BDAY:" . $vcardObject->birthday . tx_vcardmail_vcard::LINEBREAK;
		}
	}

	protected static function _renderFullname($vcardObject)
	{
		return "FN:" . $vcardObject->fullname . tx_vcardmail_vcard::LINEBREAK;
	}

	protected static function _renderName($vcardObject)
	{
		return "N:"
		. $vcardObject->lastname . ";"
		. $vcardObject->firstname . ";"
		. $vcardObject->additionalnames . ";"
		. $vcardObject->nameprefix . ";"
		. $vcardObject->namesuffix . tx_vcardmail_vcard::LINEBREAK;
	}

	protected static function _renderVersion($version)
	{
		return "VERSION:" . $version . tx_vcardmail_vcard::LINEBREAK;
	}

	protected static function _renderAddress($vcardObject)
	{

		if ($vcardObject->address)
		{
			$addressString = "";
			foreach ($vcardObject->address as $address)
			{
				$addressTypeString = '';
				if ($address['type'])
				{
					foreach ($address['type'] as $addressType)
					{
						$addressTypeString .= ";" . $addressType;
					}
				}
				$addressString .= "ADR" . $addressTypeString . ":"
								. $address['postofficeaddress'] . ";"
								. $address['extendedaddress'] . ";"
								. $address['street'] . ";"
								. $address['city'] . ";"
								. $address['state'] . ";"
								. $address['zip'] . ";"
								. $address['country'] . tx_vcardmail_vcard::LINEBREAK;
			}
			return $addressString;
		}
	}

	protected static function _renderBegin()
	{
		return "BEGIN:VCARD" . tx_vcardmail_vcard::LINEBREAK;
	}

	protected static function _renderTelephone($vcardObject)
	{
		if ($vcardObject->telephone)
		{
			$telephoneString = "";
			foreach ($vcardObject->telephone as $telephone)
			{
				$telephoneTypeString = '';

				if ($telephone['type'])
				{
					foreach ($telephone['type'] as $telephoneType)
					{
						$telephoneTypeString .= ";" . $telephoneType;
					}
				}
				$telephoneString .= "TEL" . $telephoneTypeString . ":" .
								$telephone['value'] . tx_vcardmail_vcard::LINEBREAK;
			}
			return $telephoneString;
		}
	}

	protected static function _renderEmail($vcardObject)
	{
		if ($vcardObject->email)
		{
			$emailString = "";
			foreach ($vcardObject->email as $email)
			{
				$emailTypeString = '';
				if ($email['type'])
				{
					foreach ($email['type'] as $emailType)
					{
						$emailTypeString .= ";" . $emailType;
					}
				}
				$emailString .= "EMAIL" . $emailTypeString . ":" .
								$email['value'] . tx_vcardmail_vcard::LINEBREAK;
			}
			return $emailString;
		}
	}

	protected static function _renderTitle($vcardObject)
	{
		if ($vcardObject->title)
		{
			return "TITLE:" . $vcardObject->title . tx_vcardmail_vcard::LINEBREAK;
		}
	}

	protected static function _renderRole($vcardObject)
	{
		if ($vcardObject->role)
		{
			return "ROLE:" . $vcardObject->role . tx_vcardmail_vcard::LINEBREAK;
		}
	}

	protected static function _renderOrganization($vcardObject)
	{
		if ($vcardObject->organization ||
						$vcardObject->department ||
						$vcardObject->subdepartment)
		{
			return "ORG:"
			. $vcardObject->organization . ";"
			. $vcardObject->department . ";"
			. $vcardObject->subdepartment . tx_vcardmail_vcard::LINEBREAK;
		}
	}

	protected static function _renderRevision($vcardObject)
	{
		if ($vcardObject->revision)
		{
			return "REV:" . $vcardObject->revision . tx_vcardmail_vcard::LINEBREAK;
		}
	}

	protected static function _renderUrl($vcardObject)
	{
		if ($vcardObject->url)
		{
			$urlString = "";
			foreach ($vcardObject->url as $url)
			{
				$emailTypeString = '';
				if ($url['type'])
				{
					foreach ($url['type'] as $emailType)
					{
						$emailTypeString .= ";" . $emailType;
					}
				}
				$urlString .= "URL" . $emailTypeString . ":" .
								$url['value'] . tx_vcardmail_vcard::LINEBREAK;
			}
			return $urlString;
		}
	}

	protected static function _renderUid($vcardObject)
	{
		if ($vcardObject->uid)
		{
			return "UID:" . $vcardObject->uid . tx_vcardmail_vcard::LINEBREAK;
		}
	}

	protected static function _renderGender($vcardObject)
	{
		if ($vcardObject->gender)
		{
			return "X-GENDER:" . $vcardObject->gender . tx_vcardmail_vcard::LINEBREAK;
		}
	}

	protected static function _renderNickname($vcardObject, $version)
	{
		if ($version == tx_vcardmail_vcard::VERSION30)
		{
			if ($vcardObject->nickname)
			{
				return "NICKNAME:" . $vcardObject->nickname . tx_vcardmail_vcard::LINEBREAK;
			}
		}
	}

	protected static function _renderInstantmessenger($vcardObject)
	{
		if ($vcardObject->im)
		{
			$imString = "";
			foreach ($vcardObject->im as $im)
			{
				$imTypeString = '';
				if ($im['type'])
				{
					foreach ($im['type'] as $imType)
					{
						$imTypeString .= ";" . $imType;
					}
				}
				$imString .= $im['messenger'] . $imTypeString . ":" .
								$im['value'] . tx_vcardmail_vcard::LINEBREAK;
			}
			return $imString;
		}
	}

	protected static function _renderEnd()
	{
		return "END:VCARD";
	}

	public static function generate($vcardObjects = null,
																 $version = tx_vcardmail_vcard::VERSION30)
	{
		if (!is_array($vcardObjects))
		{
			$vcardObjects = array($vcardObjects);
		}

		$vcardString = "";

		foreach ($vcardObjects as $vcardObject)
		{
			if (!empty($vcardString))
			{
				$vcardString .= tx_vcardmail_vcard::LINEBREAK;
			}
			$vcardString .= self::_renderBegin();
			$vcardString .= self::_renderVersion($version);
			$vcardString .= self::_renderFullname($vcardObject);
			$vcardString .= self::_renderName($vcardObject);
			$vcardString .= self::_renderBirthday($vcardObject);
			$vcardString .= self::_renderAddress($vcardObject);
			$vcardString .= self::_renderTelephone($vcardObject);
			$vcardString .= self::_renderEmail($vcardObject);
			$vcardString .= self::_renderInstantmessenger($vcardObject);
			$vcardString .= self::_renderTitle($vcardObject);
			$vcardString .= self::_renderRole($vcardObject);
			$vcardString .= self::_renderOrganization($vcardObject);
			$vcardString .= self::_renderRevision($vcardObject);
			$vcardString .= self::_renderUrl($vcardObject);
			$vcardString .= self::_renderUid($vcardObject);
			$vcardString .= self::_renderNickname($vcardObject, $version);
			$vcardString .= self::_renderGender($vcardObject);
			$vcardString .= self::_renderGeolocation($vcardObject);
			$vcardString .= self::_renderMailer($vcardObject);
			$vcardString .= self::_renderTimezone($vcardObject);
			$vcardString .= self::_renderEnd();
		}

		return $vcardString;
	}
}


libraries/zendvcard/class.tx_vcardmail_vcard.php
<?php

/***************************************************************
 * Copyright 2010 Thomas Schaaf <Thomaschaaf@gmail.com>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * Changes by: Franz Geiger <mail_at_fx-g.de>
 * Original code is available at:
 * http://code.google.com/p/zendvcard/
 ***************************************************************/
/**
 * A represantation of a Vcard.Version
 *
 * @author Thomas Schaaf <Thomaschaaf@gmail.com>
 */
class tx_vcardmail_vcard
{
	const VERSION21 = '2.1';
	const VERSION30 = '3.0';
	const LINEBREAK = "\r\n";

	protected $_uid = null;
	protected $_title = null;
	protected $_fullname = null;
	protected $_lastname = null;
	protected $_firstname = null;
	protected $_nickname = null;
	protected $_additionalnames = null;
	protected $_nameprefix = null;
	protected $_namesuffix = null;
	protected $_birthday = null;
	protected $_organization = null;
	protected $_department = null;
	protected $_subdepartment = null;
	protected $_role = null;
	protected $_revision = null;
	protected $_geolocation = null;
	protected $_mailer = null;
	protected $_timezone = null;
	/**
	 * @var array array(postofficeaddress, extendedaddress, street, city, state, zip, country, type=array())
	 */
	protected $_address = array();
	const ADDRESSDOMESTIC = 'DOM';
	const ADDRESSINTERNATIONAL = 'INTL';
	const ADDRESSSPOSTAL = 'POSTAL';
	const ADDRESSPARCEL = 'PARCEL';
	const ADDRESSHOME = 'HOME';
	const ADDRESSWORK = 'WORK';
	protected $addressConstantArray = array(self::ADDRESSDOMESTIC,
			self::ADDRESSHOME,
			self::ADDRESSINTERNATIONAL,
			self::ADDRESSPARCEL,
			self::ADDRESSSPOSTAL,
			self::ADDRESSWORK);
	/**
	 * @var array array(value, type=array())
	 */
	protected $_telephone = array();
	const TELEPHONEPREF = 'PREF';
	const TELEPHONEWORK = 'WORK';
	const TELEPHONEHOME = 'HOME';
	const TELEPHONEOTHER = 'OTHER';
	const TELEPHONEVOICE = 'VOICE';
	const TELEPHONEFAX = 'FAX';
	const TELEPHONEMSG = 'MSG';
	const TELEPHONECELL = 'CELL';
	const TELEPHONECAR = 'CAR';
	protected $telephoneConstantArray = array(self::TELEPHONECAR,
			self::TELEPHONECELL,
			self::TELEPHONEFAX,
			self::TELEPHONEHOME,
			self::TELEPHONEMSG,
			self::TELEPHONEOTHER,
			self::TELEPHONEPREF,
			self::TELEPHONEVOICE,
			self::TELEPHONEWORK);
	/**
	 * @var array array(value, type=array())
	 */
	protected $_email = array();
	const EMAILINTERNET = 'INTERNET';
	const EMAILWORK= 'WORK';
	const EMAILHOME = 'HOME';
	const EMAILOTHER = 'OTHER';
	const EMAILPREF = 'PREF';
	protected $emailConstantArray = array(self::EMAILHOME,
			self::EMAILINTERNET,
			self::EMAILOTHER,
			self::EMAILPREF,
			self::EMAILWORK);
	/**
	 * @var array array(value, type(array())
	 */
	protected $_url = array();
	const URLWORK= 'WORK';
	const URLHOME = 'HOME';
	const URLOTHER = 'OTHER';
	const URLPREF = 'PREF';
	protected $urlConstantArray = array(self::URLHOME,
			self::URLOTHER,
			self::URLPREF,
			self::URLWORK);
	/**
	 * @var array array(value, type=array(), $messenger=null)
	 */
	protected $_im = array();
	const IMHOME = 'HOME';
	const IMWORK = 'WORK';
	const IMPREF = 'PREF';
	const IMOTHER = 'OTHER';
	const IMAIM = 'X-AIM';
	const IMICQ = 'X-ICQ';
	const IMJABBER = 'X-JABBER';
	const IMMSN = 'X-MSN';
	const IMYAHOO = 'X-YAHOO';
	const IMSKYPE = 'X-SKYPE';
	const IMSKYPEALTERNATIVE = 'X-SKYPE-USERNAME';
	const IMGADUGADU = 'X-GADUGADU';
	protected $imMessengerConstantArray = array(
			self::IMAIM,
			self::IMICQ,
			self::IMJABBER,
			self::IMMSN,
			self::IMYAHOO,
			self::IMSKYPE,
			self::IMSKYPEALTERNATIVE,
			self::IMGADUGADU
	);
	protected $imConstantArray = array(self::IMHOME,
			self::IMOTHER,
			self::IMPREF,
			self::IMWORK);
	protected $_gender = null;
	const GENDERMALE = 'Male';
	const GENDERFEMALE = 'Female';

	public function __call($function, $args)
	{
		$function = strtolower($function);
		$property = substr($function, 3);

		if (substr($function, 0, 3) == 'set' &&
						property_exists($this, "_" . $property) &&
						!is_array($this->{"_" . $property}))
		{
			if (is_string($args[0]) || is_int($args[0]))
			{
				$this->$property = $args[0];

				return $this;
			}
		}
		else
		{
			throw new Exception("Unknown property $property");
		}
	}

	public function __get($name)
	{
		if (property_exists($this, "_" . $name))
		{
			return $this->{"_" . $name};
		}
		else
		{
			throw new Exception("Unknown property $name");
		}
	}

	public function __set($name, $value)
	{
		if (property_exists($this, "_" . $name) && !is_array($this->{"_" . $name}))
		{
			$this->{"_" . $name} = $value;
		}
		else
		{
			throw new Exception("Unknown property $name");
		}
	}

	public function addPhonenumber($phonenumberArray, $type = null)
	{
		if (!is_array($phonenumberArray))
		{
			$phonenumberCacheArray = array(array("value" => $phonenumberArray));

			if (is_array($type))
			{
				$phonenumberCacheArray[0]['type'] = $type;
			}
			else
			{
				$phonenumberCacheArray[0]['type'][] = $type;
			}

			$phonenumberArray = $phonenumberCacheArray;
		}

		foreach ($phonenumberArray as $value)
		{
			$typeArray = array();
			foreach ($value['type'] as $telephoneType)
			{
				if (in_array($telephoneType, $this->telephoneConstantArray))
				{
					$typeArray[] = $telephoneType;
				}
			}
			$this->_telephone[] = array('value' => $value['value'], 'type' => $typeArray);
		}
		return $this;
	}

	public function addEmail($emailArray, $type = null)
	{
		if (!is_array($emailArray) && is_string($emailArray))
		{
			$emailCacheArray = array();
			$emailCacheArray[0]['value'] = $emailArray;
			if (is_array($type))
			{
				$emailCacheArray[0]['type'] = $type;
			}
			else
			{
				$emailCacheArray[0]['type'][] = $type;
			}
			$emailArray = $emailCacheArray;
		}
		foreach ($emailArray as $value)
		{
			$typeArray = array();
			foreach ($value['type'] as $emailType)
			{
				if (in_array($emailType, $this->emailConstantArray))
				{
					$typeArray[] = $emailType;
				}
			}
			$this->_email[] = array('value' => $value['value'], 'type' => $typeArray);
		}
		return $this;
	}

	public function addUrl($urlArray, $type = null)
	{
		if (!is_array($urlArray) && is_string($urlArray))
		{
			$urlCacheArray = array();
			$urlCacheArray[0]['value'] = $urlArray;
			if (is_array($type))
			{
				$urlCacheArray[0]['type'] = $type;
			}
			else
			{
				$urlCacheArray[0]['type'][] = $type;
			}
			$urlArray = $urlCacheArray;
		}
		foreach ($urlArray as $value)
		{
			$typeArray = array();
			foreach ($value['type'] as $urlType)
			{
				if (in_array($urlType, $this->urlConstantArray))
				{
					$typeArray[] = $urlType;
				}
			}
			$this->_url[] = array('value' => $value['value'], 'type' => $typeArray);
		}
		return $this;
	}

	public function addAddress($addressArray)
	{
		if (isset($addressArray['type']) || !isset($addressArray[0]))
		{
			$addressCacheArray = array();
			$addressCacheArray[] = $addressArray;
			$addressArray = $addressCacheArray;
		}
		foreach ($addressArray as $value)
		{
			$typeCacheArray = array();
			if (!isset($value['type']))
			{
				$value['type'][] = null;
			}
			elseif (!is_array($value['type']))
			{
				$typeCacheArray[] = $value['type'];

				$value['type'] = $typeCacheArray;
			}

			$typeArray = array();
			foreach ($value['type'] as $addressType)
			{
				if (in_array($addressType, $this->addressConstantArray))
				{
					$typeArray[] = $addressType;
				}
			}


			$this->_address[] = array(
					'postofficeaddress' => (isset($value['postofficeaddress']) ? $value['postofficeaddress'] : ''),
					'extendedaddress' => (isset($value['extendedaddress']) ? $value['extendedaddress'] : ''),
					'street' => (isset($value['street']) ? $value['street'] : ''),
					'city' => (isset($value['city']) ? $value['city'] : ''),
					'state' => (isset($value['state']) ? $value['state'] : ''),
					'zip' => (isset($value['zip']) ? $value['zip'] : ''),
					'country' => (isset($value['country']) ? $value['country'] : ''),
					'type' => $typeArray);
		}
		return $this;
	}

	public function addInstantmessenger($imArray, $messenger = null, $type = null)
	{
		if (!is_array($imArray) && is_string($imArray) && isset($messenger))
		{
			$imCacheArray = array();
			$imCacheArray[0]['value'] = $imArray;
			$imCacheArray[0]['messenger'] = $messenger;
			if (is_array($type))
			{
				$imCacheArray[0]['type'] = $type;
			}
			else
			{
				$imCacheArray[0]['type'][] = $type;
			}
			$imArray = $imCacheArray;
		}

		foreach ($imArray as $value)
		{
			$typeArray = array();
			foreach ($value['type'] as $imArray)
			{
				if (in_array($imArray, $this->imConstantArray))
				{
					$typeArray[] = $imArray;
				}
			}
			if (in_array($value['messenger'], $this->imMessengerConstantArray))
			{
				$this->_im[] = array('value' => $value['value'], 'type' => $typeArray, 'messenger' => $value['messenger']);
			}
		}
		return $this;
	}
}
